enterprise/server/crypter_service/crypter_service.go

```diff
@@ -629,6 +641,164 @@ func (c *Crypter) NewDecryptor(ctx context.Context, digest *repb.Digest, r io.Re
+	bbmk, err := c.kms.FetchMasterKey()
+	if err != nil {
+		return err
+	}
+
+	gmk, err := c.kms.FetchKey(ekv.GroupKeyURI)
+	if err != nil {
+		return err
+	}
+
+	masterKeyPortion, err := bbmk.Decrypt(ekv.MasterEncryptedKey, nil)
+	if err != nil {
+		return err
+	}
+	groupKeyPortion, err := gmk.Decrypt(ekv.GroupEncryptedKey, []byte(ekv.GroupID))
+	if err != nil {
+		return err
+	}
+	encMasterKeyPortion, err := bbmk.Encrypt(masterKeyPortion, nil)
+	if err != nil {
+		return err
+	}
+	encGroupKeyPortion, err := gmk.Encrypt(groupKeyPortion, []byte(ekv.GroupID))
+	if err != nil {
+		return err
+	}
+
+	q := `
+		UPDATE "EncryptionKeyVersions"
+		SET master_encrypted_key = ?,
+			group_encrypted_key = ?,
+			last_encryption_attempt_at_usec = ?,
+			last_encrypted_at_usec = ?
+		WHERE encryption_key_id = ? AND version = ?
+	`
+	now := c.clock.Now()
+	args := []interface{}{encMasterKeyPortion, encGroupKeyPortion, now.UnixMicro(), now.UnixMicro(), ekv.EncryptionKeyID, ekv.Version}
+	if err := c.dbh.DB(ctx).Exec(q, args...).Error; err != nil {
+		return err
+	}
+
+	log.Infof("Successfully re-encrypted key %q version %d", ekv.EncryptionKeyID, ekv.Version)
+
+	return nil
+}
+
+func (c *Crypter) keyRencryptorIteration(cutoff time.Time) error {
+	queryKeys := func() ([]*encryptionKeyVersionWithGroupID, error) {
+		ctx, cancel := context.WithTimeout(c.env.GetServerContext(), keyReencryptListQueryTimeout)
+		defer cancel()
+		q := `
+			SELECT ek.group_id, ekv.*
+			FROM "EncryptionKeyVersions" ekv
+			JOIN "EncryptionKeys" ek ON ek.encryption_key_id = ekv.encryption_key_id
+			WHERE ekv.last_encryption_attempt_at_usec < ?
+			LIMIT 1000
+	`
+
+		retrier := retry.DefaultWithContext(ctx)
+		var lastErr error
+		var ekvs []*encryptionKeyVersionWithGroupID
+		for retrier.Next() {
+			ekvs = nil
+			rows, err := c.dbh.DB(ctx).Raw(q, cutoff.UnixMicro()).Rows()
+			if err != nil {
+				lastErr = err
+				continue
+			}
+
+			for rows.Next() {
+				var ekv encryptionKeyVersionWithGroupID
+				if err := c.dbh.DB(ctx).ScanRows(rows, &ekv); err != nil {
+					return nil, err
+				}
+				ekvs = append(ekvs, &ekv)
+			}
+			return ekvs, nil
+		}
+
+		return nil, lastErr
+	}
+
+	reencryptKey := func(ekv *encryptionKeyVersionWithGroupID) {
+		qCtx, qCancel := context.WithTimeout(c.env.GetServerContext(), keyReencryptTimeout)
+		defer qCancel()
+		retrier := retry.DefaultWithContext(qCtx)
+		for retrier.Next() {
+			if err := c.reencryptKey(qCtx, ekv); err != nil {
+				log.Warningf("could not reencrypt key %q: %s", ekv.EncryptionKeyID, err)
+			} else {
+				break
+			}
+		}
+
+		// Use a separate context in case we already fully used up the time on
+		// the previous one. We still want to make sure we have time to update
+		// the DB.
+		uCtx, uCancel := context.WithTimeout(c.env.GetServerContext(), keyReencryptTimeout/4)
+		defer uCancel()
+		// Unconditionally update the last attempt timestamp, whether we
+		// succeeded or not.
+		q := `
```

Comment:
Should we just not do this write if the re-encryption succeeded?

Reply:
good point, done